### ISSUE TITLE
ACS: RequestID + InResponseTo

### ISSUE THREAD

**Issue by ahjohannessen** [onelogin/java-saml#205]
I am confused about where I am supposed store RequestIds to use with auth.processResponse() - If 100 users try to log in at the same time, I have a cache with 100 keys. processResponse seems to assume that only user can log in at the same time, or requestId is provided by a cookie or relay state.

**Comment by pitbulk** [comment]
You can store them where you want, maybe the simple way is in a session var.

**Comment by ahjohannessen** [comment]
So, I need to establish a temporary session cookie for each user that tries to login?

Why not allow for a function exists: in-resonse-to => boolean or similar?

passport-saml uses something like a cache provider.

**Comment by pitbulk** [comment]
Can you share link to the code so I can try understand?

**Comment by ahjohannessen** [comment]
In Scala it would be something like:
```scala
def (exists: String => Boolean)
...
val response = processResponse(cache.exists)
```

**Comment by ahjohannessen** [comment]
Sure: https://www.npmjs.com/package/passport-saml

**Comment by pitbulk** [comment]
Ok, I see that Node toolkit allows defining a cacheProvider, so the processResponse method will search on the cache if the exists parameter is true.

Honestly, adding this cache stuff to the toolkit will turn it more complex.

What is the problem of storing the ID on the cookie?  other toolkits like [simpleSAMLphp stores it on the session](https://github.com/simplesamlphp/simplesamlphp/blob/3b4178d6e25031d84ff224711dd57a6e93584462/modules/saml/lib/IdP/SAML2.php#L426), and you can store the session where you want.

**Comment by ahjohannessen** [comment]
If I can get away with storing the request id in a cookie I’ll do that. Thanks. As far as I can tell, that check is just to prevent Idp initiated login?

**Comment by ahjohannessen** [comment]
The difference is that I cannot know, in the case of a global cache, which of the 100s of stored RequestId's that I am supposed to provide to `processResponse`, but in the case of a predicate handed to the method (or an alternative) it has knowledge of the `InResponseTo` value and can invoke the predicate. This works for scenarios where the RequestId comes from a cookie / session or a global cache.

**Comment by ahjohannessen** [comment]
I suppose when you switch to java 8 only, it could be as simple as:

```java
Predicate<Boolean> exists = ???
processResponse(exists);
```
and inside `processResponse` somewhere, you invoke `exists.test(id)`

**Comment by ahjohannessen** [comment]
In the case of you already have the `RequestId` you just provide `id -> (id == rid)`.

**Comment by mauromol** [comment]
IMHO driving the orchestration of the login process on SP-side at session-level should be out-of-scope for java-saml. I believe the whole subject is complex. When implementing a SPID SP, I set up a whole authentication process which keeps tracks of all created requests, matches responses with the corresponding requests, prevents replay attacks, ensures proper expiration checks, performs proper response validation and outcome processing and so on, which goes well beyond a simple "keep track of the AuthnRequest ID in session to be checked on response process".
Or perhaps a different project, built on top of java-saml, may provide such a mechanism. I see the bare java-saml more as a "toolkit", rather than a complete SAML implementation.

But perhaps @pitbulk has a different view on this.

This said, getting to the last comments by @ahjohannessen, if I understand well the need, he would like the toolkit not to perform a "when processing the response, check that the inResponseTo is equal to this request ID", but rather that "when processing the response, check that inResponseTo is among a set of request IDs I know are valid" (which check would be made through a proper `Predicate` implementation that queries a global cache). IMHO this approach is wrong: the check should not just ensure that the response is targeted to some request that was previously made, but rather the SAML consumer should determine in advance in some way EXACTLY which request the received response is targeted to and this is usually made using a session-based or some kind of relay state mechanism. This because the consumer should be certain that the `SamlResponse` is carried by the same user agent that had carried the `AuthnRequest` over the IdP (or at least that it's the one it's supposed to be): it should not just authenticate someone based on the fact that a previous request was made (possibly by someone else). I believe there are serious security implications in such a lenient approach. Correct me if I'm wrong.

**Comment by Mr-Chehata** [comment]
yes like @mauromol said,  Storing Requests IDs in global cache only prevent from  => forged response bu  does not prevent from stollen one ...

We must save the state for that specific user agent when initiating the request and make sure that response associated to that state (and that will possible only by session/cookies)
